### DISCUSSION TITLE
On configuration

### DISCUSSION THREAD

**Discussion by TheRealGramdalf** [kanidm/kanidm#2415] [Ideas]
In light of the current ~~alpha~~ release candidate status of Kanidm, I'm curious as to what your thoughts are on https://github.com/Marekkon5/onetagger/issues/314 - the majority of this issue concerns configuration, and how it should be implemented. 
to prevent repeating myself many times over, I ask that you take a peek at the issue and see how it might be relevant to Kanidm.

**Comment by Firstyear** [comment]
It's release candidate.

**Reply by Firstyear** [reply]
We already support configuration via files and the environment. So I think we already satisfy this.

**Reply by TheRealGramdalf** [reply]
Ah yes, forgot about that. In any case, if you wouldn't mind giving it a thorough read I'm genuinely curious if this is something you (as in the Kanidm dev team collectively) are interested in. Here's a better TL;DR in any case: https://github.com/kanidm/kanidm/discussions/2415#discussioncomment-8017992

**Reply by TheRealGramdalf** [reply]
> We already support configuration via files and the environment. So I think we already satisfy this.

True, for the most part. Whether intentional or not, I'm unable to set some configuration options via environment variables - database options, domain, and origin are the ones I've encountered so far. But that's probably for a separate issue.

**Comment by yaleman** [comment]
Kanidm has a documented structured configuration file and (largely) supports environment variables for config... rather than "I wrote a vaguely related thing somewhere else, please spend time reading it and work out what I mean" could you please ... explain what you're asking for or suggesting, specific to our project?

**Reply by TheRealGramdalf** [reply]
My apologies - perhaps I was to hasty writing the discussion. If you follow the link you'll find some notes specifically regarding Kanidm, which should answer your concerns - but in any case, here's the TL;DR

- Would Kanidm be interested in becoming compliant to an RFC of sorts, a standard method of configuration for dockerized programs that is both consistent and flexible across whatever language or format you use? (*I know that may be overly ambitious, but I think I have the majority of it mapped out*)
- Would Kanidm be interested in writing/contributing to a rust library that can be reused, such that other projects (in this case, OneTagger) could use it for configuration?

**Reply by Firstyear** [reply]
>     * Would Kanidm be interested in becoming compliant to an RFC of sorts, a standard method of configuration for dockerized programs that is both consistent and flexible across whatever language or format you use? (_I know that may be overly ambitious, but I think I have the majority of it mapped out_)

No - If we follow someone else's standard we become "locked in" to however that standard structures and defines it's values and configuration. Suddenly we may lose the ability to actually express or request configuration in a way that is human and intuitive to an admin. Then we have to do horrible hacks or work arounds to jam in "json blobs" to strings or similar. Everytime we take on an external dependency we are at risk of that dependency upending things on us (which has happened more than once in the life of this project). 

We need to be free to change our configuration in the interest of human accessibility and experience at any time. Being locked into some external standard will potentially be a barrier to that.

Sometimes there is a value in "not invented here" and doing things ourselves - we get to control the implementation to suit our precise needs, rather than morph our implementation to suit someone elses ideas. 

Additionally, our focus is on making an IAM - not a configuration library. We have enough on our plates as is without taking on more. 

> 
>     * Would Kanidm be interested in writing/contributing to a rust library that can be reused, such that other projects (in this case, OneTagger) could use it for configuration?

See above.

**Reply by yaleman** [reply]
To be honest, without any more context than "I have an idea, but no detail, do you want to use it?", I have no idea what we'd be contributing to, and *people* contribute to projects, not *projects*. We're not a funded organisation, just some people doing a thing.

If there's a flexible, well-designed, published thing that fulfil's people's needs, they'll use it; build your thing and people will use it if they find a need.

**Reply by TheRealGramdalf** [reply]
That makes perfect sense, and I understand why your answer is no. As always (and I hope this is abundantly clear), I respect your decision - you provide excellent software, free of charge, for the benefit of many people, and have no obligation to implement any specific feature, reply to an issue, or do any of what you currently do. Many thanks for the consideration in any case.
